src/Semantic/Api/Diffs.hs

```diff
@@ -81,20 +81,24 @@ diffGraph blobs = DiffTreeGraphResponse . V.fromList . toList <$> distributeFor
 
 type DiffEffects sig m = (Member (Error SomeException) sig, Member (Reader Config) sig, Member Telemetry sig, Member Distribute sig, Member Parse sig, Carrier sig m, MonadIO m)
 
-type Decorate a b = forall term diff . DiffActions term diff => Blob -> term a -> term b
 
+dotGraphDiffParsers :: Map Language (SomeParser DOTGraphDiff Loc)
+dotGraphDiffParsers = aLaCarteParsers
 
-class DOTGraphDiff diff where
-  dotGraphDiff :: (Carrier sig m, Member (Reader Config) sig) => diff Loc Loc -> m Builder
+class DiffTerms term => DOTGraphDiff term where
+  dotGraphDiff :: (Carrier sig m, Member (Reader Config) sig) => DiffFor term Loc Loc -> m Builder
```

Comment:
The new abstract parsing facilities require the constraints be phrased against the term, so we rephrase all of these classes against the term type, using `DiffTerms` as a supertype, and a type family dependency on `DiffFor` in order to ensure that `dotGraphDiff` is unambiguous despite only using `term` as a parameter to a type family.

I feel a bit like I’ve gained a level, in terms of using `-XTypeFamilyDependencies` successfully for the first time, but regardless of that, I’m also just quite happy with how this solution has turned out, since it means we don’t have to have separate abstraction facilities for diffs and terms any more.

The essential realization was the term/term pairs was the wrong place to place the abstraction; placing it on the _parser_ enables us to generate terms/term pairs from the same place with the assurance that a) the term type is the same between both, and b) the constraint `c` holds for both. And since `c term` in turn allows us to constrain the use of `DiffFor term`, all our term/diff processing functionality can share a small set of primitives for abstract parsing.

Reply:
tfw Rob gains a level

![](https://media.giphy.com/media/P7qFEMh3TLMGY/giphy.gif)